DependencyScanner: add a new extraPcmArgs field for each Swift module

nkcsgexi opened the pull request:
Building each Swift module explicitly requires dependency PCMs to be built
with the exactly same deployment target version. This means we may need to
build a Clang module multiple times with different target triples.

This patch removes the -target arguments from the reported PCM build
arguments and inserts extraPcmArgs fields to each Swift module.
swift-driver can combine the generic PCM arguments with these extra arguments
to get the command suitable for building a PCM specifically for
that loading Swift module.

nkcsgexi committed:
DependencyScanner: add a new extraPcmArgs field for each Swift module

Building each Swift module explicitly requires dependency PCMs to be built
with the exactly same deployment target version. This means we may need to
build a Clang module multiple times with different target triples.

This patch removes the -target arguments from the reported PCM build
arguments and inserts extraPcmArgs fields to each Swift module.
swift-driver can combine the generic PCM arguments with these extra arguments
to get the command suitable for building a PCM specifically for
that loading Swift module.

nkcsgexi commented:
@swift-ci please smoke test

artemcm commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1398,6 +1400,9 @@ bool InterfaceSubContextDelegateImpl::runInSubCompilerInstance(StringRef moduleN
   info.BuildArguments = BuildArgs;
   info.Hash = CacheHash;
+  info.ExtraPCMArgs = {"-Xcc", "-target", "-Xcc",
+    subInstance.getASTContext().LangOpts.Target.str()};
```
After this PR, we strip `-target <Triple>` from Clang module build commands, but not the triple that is specified inside `.siwftinterface` files for Swift modules. And it seems that here we will always set the `extraPCMArgs` to the triple inherited from the scanner invocation. 

What will happen when a module's interface specifies a target that does not match the one we specify here?

e.g. Swift module `A` depends on Clang module `B`. 
Module A's `.swiftinterface` file specifies: `-target x86_64-apple-macosx10.9`, and the scanner is invoked with: `-target x86_64-apple-macosx10.10`. Then the `extraPCMArgs` will specify 10.10, `A` will be built targeting 10.9, and it will depend on `B` built targeting 10.10?

Am I missing something or should we set this to the target triple specified in the `.swiftinterface` file, if one is present; and if one is not present, only then use the target triple inherited from the scanner invocation?

nkcsgexi replied:
Artem, we are actually using the target triple specified in the `.swiftinterface` file here (notice we are getting the LangOpts from the `subInstance`.

artemcm replied:
Ohh, I see! Thanks.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
test: remove test cases for building modules explicitly

These tests are duplicates of those from swift-driver.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@artemcm This will break the swift-driver side test for end-to-end explicit module building. Should we hold on merging this?

artemcm commented:
> @artemcm This will break the swift-driver side test for end-to-end explicit module building. Should we hold on merging this?

Good call. I'll grab this PR for testing while developing the swift-driver side and will ping you once we can merge them both. 👍

nkcsgexi commented:
> Good call. I'll grab this PR for testing while developing the swift-driver side and will ping you once we can merge them both. 👍

Thank you!

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
DependenciesScanner: move -fapinotes-swift-version to extra PCM arguments

Swift interface files may specify the effective language version to use. When building
a PCM loadable for these textual interface files, we should respect the language
version. This patch moves -fapinotes-swift-version from the generic PCM
commands to the extra PCM arguments owned by each loading Swift module.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@artemcm Merged
